File: website/API/services/sync_scope.go
package services

import "sort"

type SyncScope struct {
	Mode     string   `json:"mode"`
	AgentIDs []string `json:"agent_ids"`
	agents   map[string]struct{}
}

var selectedSyncTables = map[string]struct{}{
	"agents": {}, "chat_messages": {}, "short_term_messages": {},
	"long_term_memories": {}, "base_memories": {}, "planned_messages": {},
}

func NewAllSyncScope() SyncScope {
	return SyncScope{Mode: "all", AgentIDs: []string{}, agents: map[string]struct{}{}}
}

func NewSelectedSyncScope(agentIDs []string) SyncScope {
	normalized := normalizeSelectedAgentIDs(agentIDs)
	agents := make(map[string]struct{}, len(normalized))
	for _, agentID := range normalized {
		agents[agentID] = struct{}{}
	}
	return SyncScope{Mode: "selected", AgentIDs: normalized, agents: agents}
}

func (s SyncScope) AllowsAgent(agentID string) bool {
	if s.Mode == "all" {
		return true
	}
	if s.agents == nil {
		s.agents = make(map[string]struct{}, len(s.AgentIDs))
		for _, selectedID := range s.AgentIDs {
			s.agents[selectedID] = struct{}{}
		}
	}
	_, ok := s.agents[agentID]
	return ok
}

func (s SyncScope) AllowsTable(table string) bool {
	if s.Mode == "all" {
		return true
	}
	_, ok := selectedSyncTables[table]
	return ok
}

func FilterSyncItems(scope SyncScope, table string, items []map[string]interface{}) []map[string]interface{} {
	if !scope.AllowsTable(table) {
		return []map[string]interface{}{}
	}
	if scope.Mode == "all" {
		return cloneSyncItems(items)
	}

	filtered := make([]map[string]interface{}, 0, len(items))
	for _, item := range items {
		if syncItemInScope(scope, table, item) {
			filtered = append(filtered, cloneSyncItem(item))
		}
	}
	return filtered
}

func FilterSyncTombstones(scope SyncScope, tombstones []map[string]interface{}) []map[string]interface{} {
	if scope.Mode == "all" {
		return cloneSyncItems(tombstones)
	}
	filtered := make([]map[string]interface{}, 0, len(tombstones))
	for _, tombstone := range tombstones {
		table := syncString(tombstone, "table_name", "TableName")
		if !scope.AllowsTable(table) {
			continue
		}
		var agentID string
		if table == "agents" {
			agentID = syncString(tombstone, "client_id", "ClientID")
		} else {
			agentID = syncString(tombstone, "agent_id", "AgentID")
		}
		if scope.AllowsAgent(agentID) {
			filtered = append(filtered, cloneSyncItem(tombstone))
		}
	}
	return filtered
}

func syncItemInScope(scope SyncScope, table string, item map[string]interface{}) bool {
	if table == "agents" {
		return scope.AllowsAgent(syncString(item, "client_id", "ClientID", "id", "ID"))
	}
	if !scope.AllowsAgent(syncString(item, "agent_id", "AgentID")) {
		return false
	}
	if table == "long_term_memories" || table == "base_memories" || table == "planned_messages" {
		return syncString(item, "group_id", "GroupID") == ""
	}
	return true
}

func syncString(item map[string]interface{}, keys ...string) string {
	for _, key := range keys {
		value, ok := item[key]
		if !ok || value == nil {
			continue
		}
		if text, ok := value.(string); ok {
			return text
		}
	}
	return ""
}

func cloneSyncItems(items []map[string]interface{}) []map[string]interface{} {
	result := make([]map[string]interface{}, 0, len(items))
	for _, item := range items {
		result = append(result, cloneSyncItem(item))
	}
	return result
}

func cloneSyncItem(item map[string]interface{}) map[string]interface{} {
	copy := make(map[string]interface{}, len(item))
	for key, value := range item {
		copy[key] = value
	}
	return copy
}

func SortedScopeAgentIDs(scope SyncScope) []string {
	result := append([]string(nil), scope.AgentIDs...)
	sort.Strings(result)
	return result
}

File: lib/services/sync_scope.dart
import '../models/sync_policy.dart';

const allSyncTables = <String>[
  'agents',
  'chat_messages',
  'short_term_messages',
  'group_chats',
  'group_members',
  'group_messages',
  'group_short_term',
  'group_shared_memories',
  'long_term_memories',
  'base_memories',
  'planned_messages',
  'user_profiles',
  'providers',
];

const agentSyncClosureTables = <String>[
  'agents',
  'chat_messages',
  'short_term_messages',
  'long_term_memories',
  'base_memories',
  'planned_messages',
];

class SyncScope {
  SyncScope._({required this.mode, required Set<String> agentIds})
    : agentIds = Set.unmodifiable(agentIds);

  factory SyncScope.accountPolicy(SyncPolicy policy) {
    if (policy.scopeMode == SyncScopeMode.selected) {
      return SyncScope._(
        mode: SyncScopeMode.selected,
        agentIds: policy.selectedAgentIds,
      );
    }
    return SyncScope._(mode: SyncScopeMode.all, agentIds: const {});
  }

  factory SyncScope.oneShot(Set<String> agentIds) => SyncScope._(
    mode: SyncScopeMode.selected,
    agentIds: agentIds.where((id) => id.trim().isNotEmpty).toSet(),
  );

  factory SyncScope.all() => SyncScope._(
    mode: SyncScopeMode.all,
    agentIds: const {},
  );

  final SyncScopeMode mode;
  final Set<String> agentIds;

  List<String> get tables =>
      mode == SyncScopeMode.all ? allSyncTables : agentSyncClosureTables;

  bool allowsTable(String table) => tables.contains(table);

  bool allowsAgent(String agentId) =>
      mode == SyncScopeMode.all || agentIds.contains(agentId);

  List<Map<String, dynamic>> filterRows(
    String table,
    Iterable<Map<String, dynamic>> rows,
  ) {
    if (!allowsTable(table)) return const [];
    if (mode == SyncScopeMode.all) {
      return rows.map(Map<String, dynamic>.from).toList();
    }
    return rows
        .where((row) => _rowAllowed(table, row))
        .map(Map<String, dynamic>.from)
        .toList();
  }

  List<Map<String, dynamic>> filterTombstones(
    Iterable<Map<String, dynamic>> tombstones,
  ) {
    if (mode == SyncScopeMode.all) {
      return tombstones.map(Map<String, dynamic>.from).toList();
    }
    return tombstones
        .where((tombstone) {
          final table = _string(tombstone, 'table_name', 'TableName');
          if (!allowsTable(table)) return false;
          final agentId = table == 'agents'
              ? _string(tombstone, 'client_id', 'ClientID')
              : _string(tombstone, 'agent_id', 'AgentID');
          return allowsAgent(agentId);
        })
        .map(Map<String, dynamic>.from)
        .toList();
  }

  bool _rowAllowed(String table, Map<String, dynamic> row) {
    if (table == 'agents') {
      return allowsAgent(_string(row, 'client_id', 'ClientID', 'id', 'ID'));
    }
    if (!allowsAgent(_string(row, 'agent_id', 'AgentID'))) return false;
    if (table == 'long_term_memories' ||
        table == 'base_memories' ||
        table == 'planned_messages') {
      return _string(row, 'group_id', 'GroupID').isEmpty;
    }
    return true;
  }

  static String _string(
    Map<String, dynamic> row,
    String first, [
    String? second,
    String? third,
    String? fourth,
  ]) {
    for (final key in [first, second, third, fourth]) {
      if (key == null) continue;
      final value = row[key];
      if (value != null) return value.toString();
    }
    return '';
  }
}
